File: day8/solve.py
from copy import copy

program = []

with open('input.txt') as f:
    for line in f:
        operation, argument = line.split()
        program.append((operation, int(argument)))

def run(program):
    pointer     = 0
    accumulator = 0
    seen        = set()

    while True:
        seen.add(pointer)
        operation, argument = program[pointer]
        if operation == 'acc':
            accumulator += argument
            pointer += 1
        elif operation == 'jmp':
            pointer += argument
        elif operation == 'nop':
            pointer += 1
        if pointer in seen:
            return accumulator, 1 # "infinite loop detected"
        if pointer == len(program):
            return accumulator, 0 # "program terminated normally"

# Part 1
accumulator, exit_status = run(program)
assert exit_status == 1
print(accumulator)

# Part 2
for i, (operation, argument) in enumerate(program):
    program_copy = copy(program)
    if operation == 'acc':
        continue
    elif operation == 'jmp':
        program_copy[i] = ('nop', argument)
    elif operation == 'nop':
        program_copy[i] = ('jmp', argument)
    accumulator, exit_status = run(program_copy)
    if exit_status == 0:
        print(accumulator)
        break

File: day8/solve.rb
require 'set'

program = File.foreach('input.txt').map do |line|
  operation, argument = line.split
  [operation, argument.to_i]
end

def run(program)
  pointer     = 0
  accumulator = 0
  seen        = Set.new

  loop do
    seen << pointer
    operation, argument = program[pointer]
    case operation
    when 'acc'
      accumulator += argument
      pointer += 1
    when 'jmp'
      pointer += argument
    when 'nop'
      pointer += 1
    end
    return [accumulator, 1] if seen.include? pointer
    return [accumulator, 0] if pointer == program.size
  end
end

puts run(program).first

program.each_with_index do |(operation, argument), i|
  copy = program.dup
  case operation
  when 'acc'
    next
  when 'jmp'
    copy[i] = ['nop', argument]
  when 'nop'
    copy[i] = ['jmp', argument]
  end
  accumulator, exit_status = run(copy)
  if exit_status == 0
    puts accumulator
    break
  end
end
